// @verycloud/cli/dist/lib/config.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_SERVER_URL = void 0;
exports.loadConfig = loadConfig;
exports.saveConfig = saveConfig;
exports.registerMCP = registerMCP;
const fs = __importStar(require("node:fs"));
const path = __importStar(require("node:path"));
const os = __importStar(require("node:os"));
exports.DEFAULT_SERVER_URL = "https://verycloud.io";
const CONFIG_DIR = path.join(os.homedir(), ".verycloud");
const CONFIG_FILE = path.join(CONFIG_DIR, "config.json");
const CLAUDE_CONFIG = path.join(os.homedir(), ".claude.json");
function loadConfig() {
    try {
        const data = fs.readFileSync(CONFIG_FILE, "utf-8");
        return JSON.parse(data);
    }
    catch {
        return null;
    }
}
function saveConfig(config) {
    fs.mkdirSync(CONFIG_DIR, { recursive: true });
    fs.writeFileSync(CONFIG_FILE, JSON.stringify(config, null, 2) + "\n", {
        mode: 0o600,
    });
}
/**
 * Register VeryCloud MCP server in ~/.claude.json.
 * Creates the file if it doesn't exist, merges if it does.
 */
function registerMCP() {
    const mcpEntry = {
        command: "vc",
        args: ["mcp"],
    };
    try {
        let config = {};
        if (fs.existsSync(CLAUDE_CONFIG)) {
            const data = fs.readFileSync(CLAUDE_CONFIG, "utf-8");
            config = JSON.parse(data);
        }
        // Ensure mcpServers object exists
        if (!config.mcpServers || typeof config.mcpServers !== "object") {
            config.mcpServers = {};
        }
        const servers = config.mcpServers;
        // Skip if already registered
        if (servers.verycloud) {
            return false;
        }
        servers.verycloud = mcpEntry;
        config.mcpServers = servers;
        fs.writeFileSync(CLAUDE_CONFIG, JSON.stringify(config, null, 2) + "\n", { mode: 0o600 });
        return true;
    }
    catch {
        return false;
    }
}
