// @verycloud/cli/dist/lib/logger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.success = success;
exports.error = error;
exports.info = info;
exports.dim = dim;
exports.bold = bold;
exports.cyan = cyan;
exports.green = green;
exports.yellow = yellow;
exports.box = box;
exports.spinner = spinner;
const RESET = "\x1b[0m";
const BOLD = "\x1b[1m";
const DIM = "\x1b[2m";
const RED = "\x1b[31m";
const GREEN = "\x1b[32m";
const YELLOW = "\x1b[33m";
const BLUE = "\x1b[34m";
const CYAN = "\x1b[36m";
function success(msg) {
    console.log(`${GREEN}✓${RESET} ${msg}`);
}
function error(msg) {
    console.error(`${RED}✗${RESET} ${msg}`);
}
function info(msg) {
    console.log(`${BLUE}ℹ${RESET} ${msg}`);
}
function dim(msg) {
    console.log(`${DIM}${msg}${RESET}`);
}
function bold(msg) {
    return `${BOLD}${msg}${RESET}`;
}
function cyan(msg) {
    return `${CYAN}${msg}${RESET}`;
}
function green(msg) {
    return `${GREEN}${msg}${RESET}`;
}
function yellow(msg) {
    return `${YELLOW}${msg}${RESET}`;
}
function box(lines) {
    const entries = Object.entries(lines);
    const maxKey = Math.max(...entries.map(([k]) => k.length));
    const maxVal = Math.max(...entries.map(([, v]) => v.length));
    const width = maxKey + maxVal + 5;
    console.log(`┌${"─".repeat(width)}┐`);
    for (const [key, val] of entries) {
        const paddedKey = key.padEnd(maxKey);
        const paddedVal = val.padEnd(maxVal);
        console.log(`│ ${BOLD}${paddedKey}${RESET}  ${paddedVal} │`);
    }
    console.log(`└${"─".repeat(width)}┘`);
}
function spinner(msg) {
    const frames = ["⠋", "⠙", "⠹", "⠸", "⠼", "⠴", "⠦", "⠧", "⠇", "⠏"];
    let i = 0;
    const id = setInterval(() => {
        process.stdout.write(`\r${CYAN}${frames[i]}${RESET} ${msg}`);
        i = (i + 1) % frames.length;
    }, 80);
    return {
        stop(finalMsg) {
            clearInterval(id);
            process.stdout.write("\r" + " ".repeat(msg.length + 4) + "\r");
            if (finalMsg) {
                success(finalMsg);
            }
        },
    };
}
