// @verycloud/cli/dist/mcp/server.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.startMCPServer = startMCPServer;
const readline = __importStar(require("node:readline"));
const config_js_1 = require("../lib/config.js");
const api_js_1 = require("../lib/api.js");
const detector_js_1 = require("../lib/detector.js");
const app_js_1 = require("../lib/app.js");
const deploy_js_1 = require("../lib/deploy.js");
const TOOLS = [
    {
        name: "verycloud_deploy",
        description: "Deploy a project to VeryCloud. Analyzes the current directory, creates an app, provisions databases, and deploys. Returns the live URL.",
        inputSchema: {
            type: "object",
            properties: {
                name: {
                    type: "string",
                    description: "App name (optional, defaults to directory name)",
                },
                git_url: {
                    type: "string",
                    description: "Git repository URL (optional, auto-detected from git remote)",
                },
                framework: {
                    type: "string",
                    description: "Framework (optional, auto-detected). One of: nextjs, nuxt, express, nestjs, go, django, fastapi, flask, rails, docker",
                },
                database: {
                    type: "string",
                    description: "Database type to provision (optional, auto-detected). One of: postgres, mysql, redis, mongodb",
                },
            },
        },
    },
    {
        name: "verycloud_status",
        description: "Get the status of a VeryCloud app",
        inputSchema: {
            type: "object",
            properties: {
                name: {
                    type: "string",
                    description: "App name (optional, defaults to current directory name)",
                },
            },
        },
    },
    {
        name: "verycloud_logs",
        description: "Get deployment logs for a VeryCloud app",
        inputSchema: {
            type: "object",
            properties: {
                name: {
                    type: "string",
                    description: "App name (optional, defaults to current directory name)",
                },
            },
        },
    },
    {
        name: "verycloud_apps",
        description: "List all VeryCloud apps",
        inputSchema: {
            type: "object",
            properties: {},
        },
    },
];
function send(msg) {
    const json = JSON.stringify(msg);
    process.stdout.write(json + "\n");
}
function sendResult(id, result) {
    send({ jsonrpc: "2.0", id, result });
}
function sendError(id, code, message) {
    send({ jsonrpc: "2.0", id, error: { code, message } });
}
async function handleToolCall(name, args) {
    const config = (0, config_js_1.loadConfig)();
    if (!config) {
        throw new Error("Not logged in. Run 'vc login' first.");
    }
    switch (name) {
        case "verycloud_deploy": {
            const project = (0, detector_js_1.detectProject)();
            const body = {
                name: args.name || project.name,
                git_url: args.git_url || project.gitURL,
                git_branch: project.gitBranch,
                framework: args.framework || project.framework,
            };
            if (!body.git_url) {
                throw new Error("No git remote found. Initialize a git repo with a remote first.");
            }
            const dbType = args.database || project.database?.type;
            if (dbType) {
                body.database = { type: dbType };
            }
            const result = await (0, api_js_1.apiRequest)("/api/deploy/oneshot", {
                method: "POST",
                body,
            });
            // Poll for completion
            const status = await (0, deploy_js_1.waitForDeployment)(result.app.id, result.deployment.id);
            return {
                content: [
                    {
                        type: "text",
                        text: status === "running"
                            ? `Successfully deployed! App URL: ${result.url}\n` +
                                (result.service
                                    ? `Database: ${result.service.type} (${result.service.name})\n`
                                    : "") +
                                (result.database_url ? `DATABASE_URL has been set automatically.` : "")
                            : `Deployment failed. Check logs with 'vc logs ${body.name}'.`,
                    },
                ],
            };
        }
        case "verycloud_status": {
            const app = await (0, app_js_1.getAppByName)(args.name);
            if (!app) {
                return {
                    content: [{ type: "text", text: `App '${args.name || "current directory"}' not found.` }],
                };
            }
            return {
                content: [
                    {
                        type: "text",
                        text: `App: ${app.name}\nStatus: ${app.status}\nFramework: ${app.framework}\nDomain: ${app.domain}\nGit: ${app.git_url || "(none)"}\nBranch: ${app.git_branch}`,
                    },
                ],
            };
        }
        case "verycloud_logs": {
            const app = await (0, app_js_1.getAppByName)(args.name);
            if (!app) {
                return {
                    content: [{ type: "text", text: `App '${args.name || "current directory"}' not found.` }],
                };
            }
            const deployments = await (0, api_js_1.apiRequest)(`/api/apps/${app.id}/deployments`);
            if (!deployments || deployments.length === 0) {
                return {
                    content: [{ type: "text", text: "No deployments found." }],
                };
            }
            const latest = deployments[0];
            const logData = await (0, api_js_1.apiRequest)(`/api/apps/${app.id}/deployments/${latest.id}/log`);
            return {
                content: [
                    {
                        type: "text",
                        text: `Deployment v${latest.version} (${latest.status}):\n\n${logData.log || "No logs available yet."}`,
                    },
                ],
            };
        }
        case "verycloud_apps": {
            const apps = await (0, api_js_1.apiRequest)("/api/apps");
            if (!apps || apps.length === 0) {
                return {
                    content: [{ type: "text", text: "No apps found." }],
                };
            }
            const lines = apps.map((a) => `${a.name} | ${a.status} | ${a.framework} | ${a.domain}`);
            return {
                content: [
                    {
                        type: "text",
                        text: `NAME | STATUS | FRAMEWORK | DOMAIN\n${"-".repeat(60)}\n${lines.join("\n")}`,
                    },
                ],
            };
        }
        default:
            throw new Error(`Unknown tool: ${name}`);
    }
}
async function handleMessage(msg) {
    try {
        switch (msg.method) {
            case "initialize":
                sendResult(msg.id, {
                    protocolVersion: "2024-11-05",
                    capabilities: {
                        tools: {},
                    },
                    serverInfo: {
                        name: "verycloud",
                        version: "0.1.0",
                    },
                });
                break;
            case "notifications/initialized":
                // Client acknowledged initialization — no response needed
                break;
            case "tools/list":
                sendResult(msg.id, { tools: TOOLS });
                break;
            case "tools/call": {
                const params = msg.params;
                const result = await handleToolCall(params.name, params.arguments || {});
                sendResult(msg.id, result);
                break;
            }
            default:
                sendError(msg.id, -32601, `Method not found: ${msg.method}`);
        }
    }
    catch (err) {
        // Sanitize error message to avoid leaking sensitive info
        let errMsg = err.message || "Unknown error";
        errMsg = errMsg.replace(/Bearer\s+\S+/gi, "Bearer [REDACTED]");
        errMsg = errMsg.replace(/X-API-Key:\s*\S+/gi, "X-API-Key: [REDACTED]");
        errMsg = errMsg.replace(/\/Users\/[^\s]+/g, "[PATH]");
        errMsg = errMsg.replace(/\/home\/[^\s]+/g, "[PATH]");
        sendError(msg.id, -32000, errMsg);
    }
}
function startMCPServer() {
    const rl = readline.createInterface({
        input: process.stdin,
        terminal: false,
    });
    rl.on("line", async (line) => {
        try {
            const msg = JSON.parse(line);
            await handleMessage(msg);
        }
        catch {
            // Ignore unparseable lines
        }
    });
    rl.on("close", () => {
        process.exit(0);
    });
}
