// @verycloud/cli/dist/commands/deploy.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.deployCommand = deployCommand;
const readline = __importStar(require("node:readline"));
const config_js_1 = require("../lib/config.js");
const api_js_1 = require("../lib/api.js");
const detector_js_1 = require("../lib/detector.js");
const deploy_js_1 = require("../lib/deploy.js");
const logger = __importStar(require("../lib/logger.js"));
function confirm(question) {
    const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout,
    });
    return new Promise((resolve) => {
        rl.question(question, (answer) => {
            rl.close();
            resolve(answer.toLowerCase() !== "n");
        });
    });
}
async function deployCommand(options) {
    const config = (0, config_js_1.loadConfig)();
    if (!config) {
        logger.error("Not logged in. Run 'vc login' first.");
        process.exit(1);
    }
    // Detect project
    const project = (0, detector_js_1.detectProject)();
    if (!project.gitURL) {
        logger.error("No git remote found. Initialize a git repo with a remote first.");
        process.exit(1);
    }
    const appName = options.name || project.name;
    // Show summary
    const summary = {
        App: appName,
        Git: project.gitURL.replace(/https?:\/\//, ""),
        Branch: project.gitBranch,
        Framework: project.framework,
    };
    if (project.database) {
        summary["Database"] = `${project.database.type} (auto-detected)`;
    }
    console.log();
    logger.box(summary);
    console.log();
    // Confirm
    if (!options.yes) {
        const ok = await confirm("Deploy? (Y/n) ");
        if (!ok) {
            logger.info("Cancelled.");
            return;
        }
    }
    const spin = logger.spinner("Deploying...");
    try {
        const body = {
            name: appName,
            git_url: project.gitURL,
            git_branch: project.gitBranch,
            framework: project.framework,
        };
        if (project.database) {
            body.database = project.database;
        }
        const result = await (0, api_js_1.apiRequest)("/api/deploy/oneshot", {
            method: "POST",
            body,
        });
        // Poll deployment status
        const appId = result.app.id;
        const depId = result.deployment.id;
        const status = await (0, deploy_js_1.waitForDeployment)(appId, depId);
        spin.stop();
        if (status === "running") {
            console.log();
            logger.success(`Deployed! ${logger.cyan(result.url)}`);
            if (result.service) {
                logger.info(`Database: ${result.service.type} (${result.service.name})`);
            }
            if (result.database_url) {
                logger.dim(`DATABASE_URL set automatically`);
            }
        }
        else {
            logger.error("Deployment failed. Check logs with 'vc logs " + appName + "'");
            process.exit(1);
        }
    }
    catch (err) {
        spin.stop();
        logger.error(`Deploy failed: ${err.message}`);
        process.exit(1);
    }
}
