// @verycloud/cli/dist/commands/status.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.statusCommand = statusCommand;
const config_js_1 = require("../lib/config.js");
const app_js_1 = require("../lib/app.js");
const logger = __importStar(require("../lib/logger.js"));
async function statusCommand(appName) {
    const config = (0, config_js_1.loadConfig)();
    if (!config) {
        logger.error("Not logged in. Run 'vc login' first.");
        process.exit(1);
    }
    try {
        const app = await (0, app_js_1.getAppByName)(appName);
        if (!app) {
            logger.error(`App '${appName || "current directory"}' not found`);
            process.exit(1);
        }
        const statusIcon = app.status === "running"
            ? logger.green("● running")
            : app.status === "building"
                ? logger.yellow("◐ building")
                : "○ " + app.status;
        console.log();
        logger.box({
            App: app.name,
            Status: statusIcon,
            Framework: app.framework,
            Domain: app.domain,
            Git: app.git_url?.replace(/https?:\/\//, "") || "(none)",
            Branch: app.git_branch,
        });
        console.log();
    }
    catch (err) {
        logger.error(`Failed: ${err.message}`);
        process.exit(1);
    }
}
